### PR TITLE
8337664: Distrust TLS server certificates issued after Oct 2024 and anchored by Entrust Root CAs

### PR DESCRIPTION
**Pull request by franferrax**
Hi, here is a [JDK-8337664: Distrust TLS server certificates issued after Oct 2024 and anchored by Entrust Root CAs](https://bugs.openjdk.org/browse/JDK-8337664) backport, based on openjdk/jdk11u#95.

After adjusting the file paths from 11u to 8u, the backport isn't clean, but conflicts are minimal. These include a copyright line and minor `java.security-<platform>` context mismatches. You can verify this comparing 00beb507c85f335e23e51b025c16fa6940a92262 against openjdk/jdk11u@90ad5b18de314faca19e322bb21a4c33cec54785.

On top of that, the code still needed adjustments for the 8u codebase, which were addressed in a separate commit, 53e8134702c5967ffd886b8530ee5728907cae91. I made these adjustments in line with 68e393c051d3b5ed5b490f362c7ba97c75761ad8, the 8u backport of [JDK-8207258: Distrust TLS server certificates anchored by Symantec Root CAs](https://bugs.openjdk.org/browse/JDK-8207258).

#### Testing

I run `jdk/tier1` and all the tests under [`jdk/test/sun/security/ssl`](https://github.com/openjdk/jdk8u/tree/e32d62e2a39510f643b32d615e76f0ff9be3d9f3/jdk/test/sun/security/ssl), using 64-bit _slowdebug_ and _release_ images, locally built in _Fedora Linux 40_. Please note that this includes the new `X509TrustManagerImpl/Entrust/Distrust.java`, which I've also made fail by temporarily undoing the `java.security-linux` changes. I haven't found any regression against `master` (currently e32d62e2a39510f643b32d615e76f0ff9be3d9f3).

Regarding the failures in GitHub Actions, we can see that this also occurred in recent `jdk8u-dev` pull requests. For example:

* `security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java#sslrooteccca`
  `security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java#sslrootevrsaca`
    * Failed in this PR, in [_Linux x64 (jdk/tier1)_](https://github.com/franferrax/jdk8u/actions/runs/10800436167/job/29958948355), [_Linux x86 (jdk/tier1)_](https://github.com/franferrax/jdk8u/actions/runs/10800436167/job/29959071363), [_Windows x64 (jdk/tier1)_](https://github.com/franferrax/jdk8u/actions/runs/10800436167/job/29959852407) and [_Windows x86 (jdk/tier1)_](https://github.com/franferrax/jdk8u/actions/runs/10800436167/job/29959178259)
    * Same as openjdk/jdk8u-dev#430, in [_Linux x64 (jdk/tier1)_](https://github.com/ktakakuri/jdk8u-dev/actions/runs/10733047585/job/29766007028), [_Linux x86 (jdk/tier1)_](https://github.com/ktakakuri/jdk8u-dev/actions/runs/10733047585/job/29766074280), [_Windows x64 (jdk/tier1)_](https://github.com/ktakakuri/jdk8u-dev/actions/runs/10733047585/job/29766722107) and [_Windows x86 (jdk/tier1)_](https://github.com/ktakakuri/jdk8u-dev/actions/runs/10733047585/job/29766218339)

* `com/sun/jdi/PrivateTransportTest.sh`
  `security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java#sslrooteccca`
  `security/infra/java/security/cert/CertPathValidator/certification/CAInterop.java#sslrootevrsaca`
  `sun/misc/CopyMemory.java`
    * Failed in this PR, in [_macOS x64 (jdk/tier1)_](https://github.com/franferrax/jdk8u/actions/runs/10800436167/job/29959293433)
    * Same as openjdk/jdk8u-dev#430, in [_macOS x64 (jdk/tier1)_](https://github.com/ktakakuri/jdk8u-dev/actions/runs/10733047585/job/29766091975)

* `gc/concurrentMarkSweep/CheckAllocateAndSystemGC.java`
    * Failed in this PR, in [_Linux x86 (hotspot/tier1)_](https://github.com/franferrax/jdk8u/actions/runs/10800436167/job/29959071966)
    * Same as openjdk/jdk8u-dev#430, in [_Linux x86 (hotspot/tier1)_](https://github.com/ktakakuri/jdk8u-dev/actions/runs/10733047585/job/29766074578)

* `compiler/unsafe/OpaqueAccesses.java`
    * Failed in this PR, in [_macOS x64 (hotspot/tier1)_](https://github.com/franferrax/jdk8u/actions/runs/10800436167/job/29959294007)
    * Same as openjdk/jdk8u-dev#430, in [_macOS x64 (hotspot/tier1)_](https://github.com/ktakakuri/jdk8u-dev/actions/runs/10733047585/job/29766092229)

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8339194](https://bugs.openjdk.org/browse/JDK-8339194) to be approved
- [x] [JDK-8337664](https://bugs.openjdk.org/browse/JDK-8337664) needs maintainer approval

### Issues
 * [JDK-8337664](https://bugs.openjdk.org/browse/JDK-8337664): Distrust TLS server certificates issued after Oct 2024 and anchored by Entrust Root CAs (**Enhancement** - P3 - Approved)
 * [JDK-8339194](https://bugs.openjdk.org/browse/JDK-8339194): Distrust TLS server certificates issued after Oct 2024 and anchored by Entrust Root CAs (**CSR**)

### Reviewers
 * [Severin Gehwolf](https://openjdk.org/census#sgehwolf) (@jerboaa - **Reviewer**)
 * [Andrew John Hughes](https://openjdk.org/census#andrew) (@gnu-andrew - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk8u.git pull/61/head:pull/61` \
`$ git checkout pull/61`

Update a local copy of the PR: \
`$ git checkout pull/61` \
`$ git pull https://git.openjdk.org/jdk8u.git pull/61/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 61`

View PR using the GUI difftool: \
`$ git pr show -t 61`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk8u/pull/61.diff">https://git.openjdk.org/jdk8u/pull/61.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk8u/pull/61#issuecomment-2346120671)

### CODE REVIEW

**Review by jerboaa — COMMENTED** [review]
Looks mostly fine.

File: jdk/src/share/classes/sun/security/validator/EntrustTLSPolicy.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+/**
+ * This class checks if Entrust issued TLS Server certificates should be
+ * restricted.
+ */
+final class EntrustTLSPolicy {
+
+    private static final Debug debug = Debug.getInstance("certpath");
+
+    // SHA-256 certificate fingerprints of distrusted roots
+    private static final Set<String> FINGERPRINTS = new HashSet<>(Arrays.asList(
```
[the remark this answers is not in the document]
Comment (franferrax) [reply jdk/src/share/classes/sun/security/validator/EntrustTLSPolicy.java]: Updated in e96486dc7adf8fd65a398490d30d44cfbb86df4c, please note that the `SYMANTEC_TLS` distrust policy (which I checked for this backport) does not make the `FINGERPRINTS` set immutable either:

https://github.com/openjdk/jdk8u/blob/e32d62e2a39510f643b32d615e76f0ff9be3d9f3/jdk/src/share/classes/sun/security/validator/SymantecTLSPolicy.java#L66-L70

Should we also update `SymantecTLSPolicy.java`?

[the remark this answers is not in the document]
Comment (jerboaa) [reply jdk/src/share/classes/sun/security/validator/EntrustTLSPolicy.java]: Thanks!

> Should we also update `SymantecTLSPolicy.java`?

No, not in this bug. Feel free to do this as an 8u-only fix via jdk8u-dev pr. The original backport to 8 of https://bugs.openjdk.org/browse/JDK-8207258 didn't do this. That's one of the gotchas of `Set.of()` backports. It's not terribly important.

**Review by jerboaa — APPROVED** [review]
Looks good.

**Comment by franferrax** [comment]
Adding @gnu-andrew for awareness. According to the [Crypto Roadmap](https://www.java.com/en/jre-jdk-cryptoroadmap.html), this change is targeting 8u, and planned for the October CPU. [23u](https://github.com/openjdk/jdk23u/pull/91), [21u](https://github.com/openjdk/jdk21u/pull/451), [17u](https://github.com/openjdk/jdk17u/pull/396) and [11u](https://github.com/openjdk/jdk11u/pull/95) backports are already integrated.

File: jdk/src/share/classes/sun/security/validator/EntrustTLSPolicy.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (gnu-andrew) [reply jdk/src/share/classes/sun/security/validator/EntrustTLSPolicy.java]: This is a common case, as Severin implies. For the test library, we added internal versions of the `listOf` and `setOf` methods to better handle this. It has the advantage that the same error checking (null values, duplicates) is present as on the methods in later JDKs.

I can look at moving that to an internal JDK class instead to cover library cases, and fix the Symantec case then. No reason to delay this critical fix for that.

**Review by gnu-andrew — APPROVED** [review]
Backport looks good to me. Missing collection and time methods are ported to suitable 8u equivalents. `java.security` changes are duplicated to the set of files in 8u.

I see differences with `entrustrootcag4-chain.pem` but this seems to be because 11u's version has CRLF line endings for some reason. The 8u one is actually correct in using the usual line endings.

**Comment by gnu-andrew** [comment]
> Adding @gnu-andrew for awareness. According to the [Crypto Roadmap](https://www.java.com/en/jre-jdk-cryptoroadmap.html), this change is targeting 8u, and planned for the October CPU. [23u](https://github.com/openjdk/jdk23u/pull/91), [21u](https://github.com/openjdk/jdk21u/pull/451), [17u](https://github.com/openjdk/jdk17u/pull/396) and [11u](https://github.com/openjdk/jdk11u/pull/95) backports are already integrated.

I'm aware of the fix from 21u. Backport looks good so feel free to apply for approval so we can get this in the October release.

**Comment by gnu-andrew** [comment]
I'll also take a look at the PR you reference for the tests and see if we can get that fixed in -dev. It won't change these results though.

File: jdk/src/share/classes/sun/security/validator/EntrustTLSPolicy.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (gnu-andrew) [reply jdk/src/share/classes/sun/security/validator/EntrustTLSPolicy.java]: Opened https://bugs.openjdk.org/browse/JDK-8340414

[the remark this answers is not in the document]
Comment (franferrax) [reply jdk/src/share/classes/sun/security/validator/EntrustTLSPolicy.java]: > Opened https://bugs.openjdk.org/browse/JDK-8340414

Great!

**Comment by franferrax** [comment]
> I see differences with `entrustrootcag4-chain.pem` but this seems to be because 11u's version has CRLF line endings for some reason. The 8u one is actually correct in using the usual line endings.

I hadn't noticed this, the [original version](https://github.com/openjdk/jdk/blob/bbb516163d400a9c7e923e423fe2a60091b59322/test/jdk/sun/security/ssl/X509TrustManagerImpl/Entrust/entrustrootcag4-chain.pem) has mixed CR and CRLF, and it got fixed when adjusting and applying the patch. If you prefer it to be a verbatim copy I can easily modify it to match the original.

> I'm aware of the fix from 21u. Backport looks good so feel free to apply for approval so we can get this in the October release.

Thanks, here is the approval [request](https://bugs.openjdk.org/browse/JDK-8337664?focusedId=14704789&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14704789).

**Comment by gnu-andrew** [comment]
> > I see differences with `entrustrootcag4-chain.pem` but this seems to be because 11u's version has CRLF line endings for some reason. The 8u one is actually correct in using the usual line endings.
> 
> I hadn't noticed this, the [original version](https://github.com/openjdk/jdk/blob/bbb516163d400a9c7e923e423fe2a60091b59322/test/jdk/sun/security/ssl/X509TrustManagerImpl/Entrust/entrustrootcag4-chain.pem) has mixed CR and CRLF, and it got fixed when adjusting and applying the patch. If you prefer it to be a verbatim copy I can easily modify it to match the original.

No need to change it. If anything, 11u should be fixed.

> 
> > I'm aware of the fix from 21u. Backport looks good so feel free to apply for approval so we can get this in the October release.
> 
> Thanks, here is the approval [request](https://bugs.openjdk.org/browse/JDK-8337664?focusedId=14704789&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14704789).

Thanks, I didn't see the `/approval` command for it on this PR.

/approve yes

**Comment by franferrax** [comment]
> No need to change it. If anything, 11u should be fixed.

Ok, just for the record, 17u, 21u, 23u and mainline also have the mixed line endings in `entrustrootcag4-chain.pem`.

> Thanks, I didn't see the `/approval` command for it on this PR.

Yes, I manually filled the request in the JBS because other backports requests were done in that way, and I was wondering which label the `/approval` command would add (`jdk8u-fix-request` vs `jdk8u-critical-request`).

**Pull request closed without merging**